### PR TITLE
libct/cg/v1: work around CPU quota period set failure

### PR DESCRIPTION
**Pull request by kolyshkin**
_This is based on (and currently includes) #3175._

I. As reported in issue #3084, sometimes setting CPU quota period fails
when a new period is lower and a parent cgroup has CPU quota limit set.

This happens as in cgroup v1 the quota and the period can not be set
together (this is fixed in v2), and since the period is being set first,
`new_limit = old_quota/new_period` may be higher than the parent
cgroup limit.

The fix is to retry setting the period after the quota, to cover all
possible scenarios.

Add a test case to cover a regression caused by an earlier version of
this patch (ignoring a failure of setting invalid period when quota is
not set).

Fixes: #3084
Closes: #2044

II. Add a test case for an issue fixed by the previous commit.
    
Unfortunately, this is somewhat complicated as there's no easy way to
create a transient unit, so a binary, sd-helper, had to be added. On top
of that, an ability to create a parent/pod cgroup is added to
helpers.bash, which might be useful for future integration tests.

#### History
* v2: rearranged, cleanups split to #3175, simplified sd-helper to not depend on urlfave/cli.

### CODE REVIEW

**Comment by kolyshkin** [comment]
A repro (which took me a while to write) fails as expected! 

> time="2021-07-15T01:24:25Z" level=error msg="failed to write \"10000\": write /sys/fs/cgroup/cpu,cpuacct/runc-cgroups-integration-test/pod_31883/test-cgroup-14305/cpu.cfs_period_us: invalid argument"
152

Adding the fix now.

**Comment by kolyshkin** [comment]
Don't know how to create a transient scope from bash :( and this is required for centos 7's systemd.

**Comment by kolyshkin** [comment]
> Don't know how to create a transient scope from bash :( and this is required for centos 7's systemd.

In theory, I should be able to use `dbus-send` to call `org.freedesktop.systemd1.Manager.StartTransientUnit` but I can't figure out a way to.

Another way is to use libcontainer and have a small go binary to do that.

**Comment by kolyshkin** [comment]
> Another way is to use libcontainer and have a small go binary to do that.

Trying it as I can't think of any better way right now.

**Comment by kolyshkin** [comment]
Might make sense to backport to 1.0 (maybe only the fix itself, as it's short and clear, unlike the test case).

File: libcontainer/cgroups/fs/cpu.go [hunk 1]
```diff
@@ -71,15 +73,32 @@ func (s *CpuGroup) Set(path string, r *configs.Resources) error {
...
+		period = strconv.FormatUint(r.CpuPeriod, 10)
+		if err := cgroups.WriteFile(path, "cpu.cfs_period_us", period); err != nil {
+			// Sometimes when the period to be set is smaller
+			// than the current one, it is rejected by the kernel
+			// (EINVAL) as old_quota/new_period exceeds the parent
+			// cgroup quota limit. If this happens, ignore the
+			// error for now and retry after setting the quota.
+			if !errors.Is(err, unix.EINVAL) {
+				return err
+			}
```
Comment (hk-vmg) [inline libcontainer/cgroups/fs/cpu.go:88]: If Quota is not being set (r.CpuQuota == 0), this will silently fail Period change as well. Any idea whether it's possible to get EINVAL in cases other than the original bug report?

Comment (kolyshkin) [reply libcontainer/cgroups/fs/cpu.go:88]: Good catch. Funny that I added quite a few corner cases for cpu quota (e.g. in #2458) and it's still not enough :)

Comment (kolyshkin) [reply libcontainer/cgroups/fs/cpu.go:89]: Fixed, and added yet another corner test case ("setting invalid period without quota should fail").

**Comment by kolyshkin** [comment]
Ughm; the backport (#3115) is already merged, and this one has no LGTMs :-\ @opencontainers/runc-maintainers PTAL 🙏🏻

**Comment by kolyshkin** [comment]
@opencontainers/runc-maintainers PTAL 🙏🏻

File: contrib/cmd/sd-helper/helper.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	"github.com/opencontainers/runc/libcontainer/configs"
+)
+
+// version will be populated by the Makefile, read from
+// VERSION file of the source code.
+var version = ""
+
+// gitCommit will be the hash that the binary was built from
+// and will be populated by the Makefile.
+var gitCommit = ""
```
Comment (thaJeztah) [inline contrib/cmd/sd-helper/helper.go:22]: Will this utility only be used for tests? Wondering if adding version information is a bit too much for that (or even using `urfave/cli` for this).

Reason I'm looking at the `version` variable, is also that if we want to switch to `//go:embed` (https://github.com/opencontainers/runc/pull/3163), the path to files to embed cannot traverse to parent directories (so `//go:embed ../../VERSION` won't work) 😅

**Review by thaJeztah — COMMENTED** [review]
Didn't do a full review yet, but left a comment. 

Also looks like the last commit is somewhat unrelated (could make sense separately)

**Comment by kolyshkin** [comment]
> Also looks like the last commit is somewhat unrelated (could make sense separately)

That's right, but it's harder to work on it if it's separate (since other commits in this PR also change the same file), and I don't want to forget about it.

Let me try to rework it so the shellcheck fixes are separate.

File: contrib/cmd/sd-helper/helper.go (hunk 2 above)
Comment (kolyshkin) [reply contrib/cmd/sd-helper/helper.go:22]: I just copied this bit (as well as using `urlfave/cli`) from `recvtty`.

Let me try to drop both.

Comment (kolyshkin) [reply contrib/cmd/sd-helper/helper.go]: Fixed, PTAL @thaJeztah

**Comment by kolyshkin** [comment]
> Let me try to rework it so the shellcheck fixes are separate.

Those fixes separated out to #3175 but this one depends on (and currently includes) it.

We can either review/merge #3175 first, and then I'll rebase, or merge this one as is and close #3175.

**Comment by kolyshkin** [comment]
This fixes a real bug, and the backport to 1.0 was already merged and released in 1.0.2. @opencontainers/runc-maintainers PTAL

**Review by AkihiroSuda — APPROVED** [review]

**Comment by kolyshkin** [comment]
For the reference, the same fix for crun: https://github.com/containers/crun/pull/1188
